src/input-states.js

```diff
@@ -0,0 +1,242 @@
+    var port = node.inPorts[portName];
+    if (!isMultiPort(port))
+        throw Error("This port is not addressable/multi: " + portName);
+    port.listAttached().forEach(function(socketId, i) {
+        setPortStateBySocketId(node, vnid, portName, socketId, stateArray[i]);
+    }, []);
+    return node;
+}
+
+/**
+ * Changes the state of a addressable/multi port socket. When a new state is received
+ * on a port this should be called.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName name of the port on this node
+ * @param socketId the socketId pass to the data event handler
+ * @param state the new state of this port socket
+ */
+function setPortStateBySocketId(node, vnid, portName, socketId, state) {
+    if (socketId == null)
+        return setPortState(node, vnid, portName, state);
+    else if (!isMultiPort(node.inPorts[portName]))
+        throw Error("This port is not addressable/multi: " + portName);
+    else if (_.isUndefined(state)) {
+        if (ensureVniStateExists(node, vnid, portName)) {
+            node.vnis[vnid].inputStates[portName][socketId] = state;
+            if (!_.compact(node.vnis[vnid].inputStates[portName]).length) {
+                delete node.vnis[vnid].inputStates[portName];
+                if (_.isEmpty(node.vnis[vnid].inputStates)) {
+                    delete node.vnis[vnid].inputStates;
+                    if (_.isEmpty(node.vnis[vnid])) {
+                        delete node.vnis[vnid];
+                    }
+                }
+            }
+        }
+    } else {
+        if (!ensureVniStateExists(node, vnid, portName)) {
+            node.vnis = node.vnis || {};
+            node.vnis[vnid] = node.vnis[vnid] || {};
+            node.vnis[vnid].inputStates = node.vnis[vnid].inputStates || {};
+            node.vnis[vnid].inputStates[portName] = node.vnis[vnid].inputStates[portName] || [];
+        }
+        node.vnis[vnid].inputStates[portName][socketId] = state;
+    }
+    return node;
+}
+
+/**
+ * Checks that the given object is an addressable/multi port. This is used to validate
+ * that a given portName is actually addressable/multi to fail fast otherwise.
+ * @param node a node facade
+ * @param port a port facade.
+ */
+function isMultiPort(port) {
+    return isPort(port) && port.isMulti();
+}
+
+/**
+ * Checks that the given object is a port or a port facade. Noflo does not provide
+ * a list of available port in a node and this allows the caller to remove
+ * other properties and functions that are on the inPorts objects.
+ * @param port a port facade.
+ */
+function isPort(port) {
+    return _.isFunction(port.isMulti);
+}
+
+/**
+ * Checks if a state already exists for this vnid and portName
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName name of the port on this node
+ */
+function ensureVniStateExists(node, vnid, portName) {
```

Comment:
If this is only checking to see if the state exists, then I think it should be renamed to stateExists.  When I previously suggested ensureVniStateExists, my intent is that it would *create* the state if it did not already exist.  I'm fine with it being done either way though.  The main thing is that I wanted to ensure that all state creation is done in a single function.